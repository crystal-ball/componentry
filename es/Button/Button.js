import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import elementFactory from '../component-factories/element';

/**
 * The library makes use of Buttons that should not have the base `.btn` class. It's
 * more explicit to specify this as a component type (and easier to debug/reason
 * about).
 * @param {boolean} decorated
 */
var makeButton = function makeButton(decorated) {
  return elementFactory(decorated ? 'Button' : 'BaseButton', function (_ref) {
    var _className;

    var color = _ref.color,
        link = _ref.link,
        outline = _ref.outline,
        size = _ref.size,
        props = _objectWithoutProperties(_ref, ["color", "link", "outline", "size"]);

    return _objectSpread({
      tag: 'button',
      type: 'button',
      className: (_className = {
        btn: decorated,
        'btn-anchor': link
      }, _defineProperty(_className, "btn-".concat(color), color && !link && !outline), _defineProperty(_className, "btn-outline-".concat(color), outline), _defineProperty(_className, 'btn-sm', size === 'small'), _defineProperty(_className, 'btn-lg', size === 'large'), _className)
    }, props);
  });
};

var Button = makeButton(true);
var BaseButton = makeButton(false);
export { BaseButton };
export default Button;