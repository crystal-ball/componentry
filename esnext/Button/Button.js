import elementFactory from '../component-factories/element';

/**
 * The library makes use of Buttons that should not have the base `.btn` class. It's
 * more explicit to specify this as a component type (and easier to debug/reason
 * about).
 * @param {boolean} decorated
 */
const makeButton = decorated => elementFactory(decorated ? 'Button' : 'BaseButton', ({
  color,
  link,
  outline,
  size,
  ...props
}) => ({
  tag: 'button',
  type: 'button',
  className: {
    btn: decorated,
    'btn-anchor': link,
    // btn-<COLOR> class is only for regular themed buttons, suppress for other
    // btn theme flavors
    [`btn-${color}`]: color && !link && !outline,
    [`btn-outline-${color}`]: outline,
    'btn-sm': size === 'small',
    'btn-lg': size === 'large'
  },
  ...props
}));

const Button = makeButton(true);
const BaseButton = makeButton(false);
export { BaseButton };
export default Button;