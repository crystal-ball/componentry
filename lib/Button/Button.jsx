import { bool, func, oneOf, shape, string } from 'prop-types'
import classNames from 'classnames'

import { renderContainer } from '../utils/element-factory'
import { suppressBoxShadowOnClick } from '../utils/dom'
import { themeColorNames } from '../utils/theme'

/**
 * The `Button` component is the base component for any element that has a user
 * interaction in the library. It is important to use either a button element or a
 * valid href for any click target in order to support keyboard users _(See A++
 * Accessibility Guide)_. In cases where a target that looks like a link is required,
 * but the target is causes an in page change, the `Button` component should be used
 * and passed the `link` property.
 *
 * TODO: Document that a default theme color is used for every button that is not a
 * link button, this can be suppressed by passing `color={null}` to component.
 *
 * _NOTE: internally component will call function `suppressBoxShadowOnClick` on the
 * mousedown event, which is only triggered by clicks. This will suppress the default
 * Bootstrap box shadow applied to buttons only on click. Keyboard users will still
 * benefit from the box shadow focus styles.
 * @return {Component}
 */

Button.propTypes = {
  /** Additional CSS classes */
  className: string,
  /**
   * Theme color for button, used to create themed buttons and themed outline
   * buttons. A default value for color is passed from the `ThemeProvider` as
   * `defaultButtonColor` and is used with any button that is not a `link` button.
   * Because sometimes it's desirable to have a button that is not a `link` and ALSO
   * not a themed button, it's possible to pass `null` for color and the default
   * theme value will not be used. See the `Dropdown.Item` for an example of this.
   *
   * Empty string is a valid value b/c that is the default that tells the component
   * to use the default theme color configured in the context.
   */
  color: oneOf([...themeColorNames, 'link', '']),
  /**
   * A++ Accessibility: Creates a button that looks exactly like an anchor. This
   * should be used for any action trigger in an application that is not a routing
   * event.
   */
  link: bool,
  // This is checked in props to handle merging with internal mouse down handler,
  // but usage should be invisible to consumers
  onMouseDown: func,
  /** Creates outline style button, uses `color` for outline theme. */
  outline: bool,
  /** Create a small or large style button */
  size: oneOf(['large', 'small']),
  /** Buttons have `type="button"` by default, pass a type to override. */
  type: string
}

Button.defaultProps = {
  className: null,
  color: '', // Do not default to null! null means do not use theme color at all
  link: false,
  onMouseDown() {},
  outline: false,
  size: null,
  type: 'button'
}

Button.contextTypes = {
  COMPONENTRY_THEME: shape({
    defaultButtonColor: string
  })
}

export default function Button(
  { className, color, link, outline, size, onMouseDown, ...rest },
  { COMPONENTRY_THEME: { defaultButtonColor = 'primary' } = {} }
) {
  const mouseDown = evt => {
    // Call passed mouse down event and then handle blur
    onMouseDown.call(this, evt)
    suppressBoxShadowOnClick(evt)
  }

  // Populate default element type
  rest.As = rest.As || 'button'
  // Pass null to suppress output of any theme color classes
  color = color || color === null ? color : defaultButtonColor

  // Always include class 'btn' and passed className
  // btn-color for non link, non outline buttons that have a theme color (color can be
  // suppressed by passing color=null)
  className = classNames('btn', className, {
    [`btn-${color}`]: color && !link && !outline,
    'btn-anchor': link, // Will create a button that looks just like an anchor
    [`btn-outline-${color}`]: outline,
    'btn-lg': size === 'large',
    'btn-sm': size === 'small'
  })

  return renderContainer({ className, onMouseDown: mouseDown, ...rest })
}
