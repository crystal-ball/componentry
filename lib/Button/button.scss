// ========================================================
// Buttons
// ========================================================

// Assumed class names:
// .btn, .btn-link, .bnt-<COLOR>, .btn-outline-<COLOR>

// NOTE: Bootstrap purposely doesn't have pointer for button cursor
// See: https://github.com/twbs/bootstrap/pull/21812

// Bootstrap has a 'link' button with the font styles of an anchor AND all of the
// padding of a button. For A++ Accessibility we want to have buttons that look
// exactly like links for in application triggers, `btn-anchor` creates a button
// styled exactly like an anchor. This should be used for application actions that
// are not route navigations.
.btn-anchor {
  @extend .btn-link; // Extend link styles for font styling
  // Remove additional button styles for anchor appearance
  -webkit-appearance: none !important; // Remove Chrome native button styling
  border: none;
  border-radius: 0;
  font-family: $font-family-sans-serif;
  line-height: $line-height-base;
  padding: 0;
  user-select: auto;
  vertical-align: baseline;
  white-space: normal;
}
