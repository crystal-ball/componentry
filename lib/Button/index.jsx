// @flow
import { createElement } from 'react'
import type { ComponentType, Node } from 'react'
import { shape, string } from 'prop-types'
import classNames from 'classnames'

import type { ThemeColors } from '../utils/theme'

type Props = {
  /**
   * Render component as
   */
  as?: ComponentType<any> | string,
  /**
   * Children
   */
  children?: Node,
  /**
   * Additional CSS classes
   */
  className?: string,
  /**
   * Theme color for button, used to create themed buttons and themed outline
   * buttons. A default value for color is passed from the `ThemeProvider` as
   * `defaultButtonColor` and is used with any button without a passed value for
   * color.
   *
   * If a button is not passed a theme color, it will default to the
   * `defaultButtonColor` value that can be set in the THEME context (it is
   * `'primary'` by default). If you need a button that is not a `link` and not a
   * themed button, pass an empty string to suppress both classes.
   */
  color?: ThemeColors | 'link' | '',
  /**
   * A++ Accessibility: Creates a button that looks exactly like an anchor. This
   * should be used for any action trigger in an application that is not a routing
   * event.
   */
  link?: boolean,
  /**
   * Creates outline style button, uses `color` for outline theme.
   */
  outline?: boolean,
  /**
   * Create a small or large style button
   */
  size?: 'small' | 'large'
}

type Context = {
  THEME: { defaultButtonColor: string }
}

/**
 * The `Button` component is the base component for any element that has a user
 * interaction in the library. It is important to use either a button element or a
 * valid href for any click target in order to support keyboard users _(See A++
 * Accessibility Guide)_. In cases where a target that looks like a link is
 * required, but the target causes an in page change, the `Button` component should
 * be used with the `link` prop.
 */
const Button = (
  { as, className, color, link, outline, size, children, ...rest }: Props,
  { THEME: { defaultButtonColor = 'primary' } = {} }: Context
) => {
  // If color isn't passed, use defaultButtonColor
  const renderColor: string = color === undefined ? defaultButtonColor : color

  return createElement(
    as || 'button',
    {
      type: 'button',
      // Always include class 'btn' and passed className
      // btn-color for non link, non outline buttons that have a theme color
      className: classNames('btn', className, {
        [`btn-${renderColor}`]: renderColor && !link && !outline,
        'btn-anchor': link, // Will create a button that looks just like an anchor
        [`btn-outline-${renderColor}`]: outline,
        'btn-lg': size === 'large',
        'btn-sm': size === 'small'
      }),
      ...rest
    },
    children
  )
}

Button.contextTypes = {
  THEME: shape({ defaultButtonColor: string })
}

export default Button
