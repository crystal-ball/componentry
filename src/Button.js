import React, { PropTypes } from 'react';
import classnames from 'classnames';

/**
 * Function to handle removing default Bootstrap box-shadow focus style only on click
 * of button by attaching function to the `onMouseDown` event (which is only
 * fired for mouse events).
 *
 * Function overrides default browser outline css, then attaches a blur
 * listener to element that returns outline control to browser on blur and
 * self destructs listener. This allows the element to regain focus if the
 * user switches to keyboard.
 *
 * _Note that Bootstrap overrides the default browser **outline** to none for button
 * focus and replaces it with a box shadow. This targets that box shadow for click
 * events **only**. On keyboard focus, the default Bootstrap box shadow is still
 * shown. This provides A++ Accessibility for keyboard users._
 * @param {Object} evt React synthetic event
 */
const suppressBoxShadowOnClick = evt => {
  function blurHandler(event) {
    // Remove box-shadow override to enable possibility of keyboard focus
    event.target.style.boxShadow = '';
    // Remove this blur listener
    event.target.removeEventListener('blur', blurHandler);
  }

  evt.target.style.boxShadow = 'none';
  evt.target.addEventListener('blur', blurHandler);
};

/**
 * The `Button` component is the base component for any element that has a user
 * interaction in the library. It is important to use either a button element or a
 * valid href for any click target in order to support keyboard users _(See A++
 * Accessibility Guide)_. In cases where a target that looks like a link is required,
 * but the target is causes an in page change, the `Button` component should be used
 * and passed the `link` property.
 *
 * _NOTE: internally component will call function `suppressBoxShadowOnClick` on the
 * mousedown event, which is only triggered by clicks. This will suppress the default
 * Bootstrap box shadow applied to buttons only on click. Keyboard users will still
 * benefit from the box shadow focus styles.
 * @param {Object} [children]
 * @param {string} [className='']
 * @param {string} [color='']       Pass a color for a contextualized button using color
 * @param {boolean} [large=false]   Pass true to render a size large button
 * @param {boolean} [link=false]    Pass true to render a button that looks like an
 *                                  anchor element (use for A++ Accessibility)
 * @param {?Function} [onMouseDown] Any function passed will be called on mousedown
 *                                  event
 * @param {boolean} [outline=false] Pass true to render an outline style button
 * @param {boolean} [small=false]   Pass true to render a size small button
 * @param {string} [type='button']  Pass a type to override button `type` attribute
 * @return {Component}
 */
const Button = ({
  children=null,
  className='',
  color='',
  large=false,
  link=false,
  outline=false,
  small=false,
  type='button',
  onMouseDown,
  ...other
}) => {
  let _onMouseDown;
  let _className = classnames('btn', className, {
    [`btn-${color}`]: color && !outline,
    'btn-link': link,
    'btn-unstyled': link,
    [`btn-outline-${color}`]: outline && color,
    'btn-lg': large,
    'btn-sm': small
  });

  // If an onMouseDown was passed in, call it, then call our blur handler
  if (onMouseDown) {
    _onMouseDown = function(evt) {
      onMouseDown.call(this, evt);
      suppressBoxShadowOnClick(evt);
    };
  // Otherwise just attach our blur handler
  } else {
    _onMouseDown = suppressBoxShadowOnClick;
  }

  return(
    <button className={_className} onMouseDown={_onMouseDown} type={type} {...other}>
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.string,
  large: PropTypes.bool,
  link: PropTypes.bool,
  onMouseDown: PropTypes.func,
  outline: PropTypes.bool,
  small: PropTypes.bool,
  type: PropTypes.string
};

Button.defaultProps = {
  children: null,
  className: null,
  color: '',
  large: false,
  link: false,
  onMouseDown: null,
  outline: false,
  small: false,
  type: 'button'
};

export default Button;
