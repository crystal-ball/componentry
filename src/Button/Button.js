// @flow
import { createElement } from 'react'
import classnames from 'classnames'
import withTheme from '../withTheme/withTheme'
import { type ElementProps } from '../component-factories/element'
import type { ThemeColors } from '../utils/theme'

export type Props = {
  /** Creates a full-width button */
  block?: boolean,
  /** Theme color used to compute BS color class */
  color?: ThemeColors | 'link' | '',
  /**
   * Semi-private internal prop that determines if component is deocrated with
   * base `btn` class
   */
  decorated: boolean,
  /** Computes the button as anchor class */
  link?: boolean,
  /** Creates outline style button, uses `color` for outline theme. */
  outline?: boolean,
  /** Create a small or large style button */
  size?: 'small' | 'large',
} & ElementProps

const Button = ({
  as,
  children,
  className,
  decorated,
  block,
  color,
  link,
  outline,
  size,
  ...rest
}: Props) =>
  createElement(
    as || 'button',
    {
      type: 'button',
      className: classnames(className, {
        btn: decorated,
        'btn-anchor': link,
        'btn-block': block,
        // btn-<COLOR> class is only for regular themed buttons, suppress for other
        // btn theme flavors
        [`btn-${color}`]: color && !link && !outline,
        [`btn-outline-${color}`]: outline,
        'btn-sm': size === 'small',
        'btn-lg': size === 'large',
      }),
      ...rest,
    },
    children,
  )
Button.displayName = 'Button'
Button.defaultProps = {
  decorated: true,
}
export default withTheme(Button)
