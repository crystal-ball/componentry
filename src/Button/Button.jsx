// @flow
import elementFactory, { type ElementProps } from '../component-factories/element'
import type { ThemeColors } from '../utils/theme'

export type Props = {
  /** Theme color used to compute BS color class */
  color?: ThemeColors | 'link' | '',
  /** Computes the button as anchor class */
  link?: boolean,
  /** Creates outline style button, uses `color` for outline theme. */
  outline?: boolean,
  /** Create a small or large style button */
  size?: 'small' | 'large',
} & ElementProps

/**
 * The library makes use of Buttons that should not have the base `.btn` class. It's
 * more explicit to specify this as a component type (and easier to debug/reason
 * about).
 * @param {boolean} decorated
 */
const makeButton = decorated =>
  elementFactory(
    decorated ? 'Button' : 'BaseButton',
    ({ color, link, outline, size, ...props }) => ({
      tag: 'button',
      type: 'button',
      className: {
        btn: decorated,
        'btn-anchor': link,
        // btn-<COLOR> class is only for regular themed buttons, suppress for other
        // btn theme flavors
        [`btn-${color}`]: color && !link && !outline,
        [`btn-outline-${color}`]: outline,
        'btn-sm': size === 'small',
        'btn-lg': size === 'large',
      },
      ...props,
    }),
  )

const Button = makeButton(true)
const BaseButton = makeButton(false)

export { BaseButton }
export default Button
