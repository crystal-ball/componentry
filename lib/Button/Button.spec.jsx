import React from 'react'
import { shallow } from 'enzyme'

import Button from './index'

describe('<Button/>', () => {
  test('should render a button element default', () => {
    const wrapper = shallow(<Button />)
    expect(wrapper.find('button').length).toEqual(1)
    expect(wrapper.find('.btn').length).toEqual(1)
    expect(wrapper.find('button[type="button"]').length).toEqual(1)
  })

  test('should render any brand classes for prop color', () => {
    const wrapper = shallow(<Button color="info" />)
    expect(wrapper.find('.btn.btn-info').length).toEqual(1)
  })

  test('should render outline class for prop outline', () => {
    const wrapper = shallow(<Button color="primary" outline />)
    // Even though color is passed, primary flag should override
    expect(wrapper.find('.btn.btn-primary').length).toEqual(0)
    expect(wrapper.find('.btn.btn-outline-primary').length).toEqual(1)
  })

  test('should render class for small button', () => {
    const wrapper = shallow(<Button size="small" />)
    expect(wrapper.find('.btn.btn-sm').length).toEqual(1)
  })

  test('should render class for large button', () => {
    const wrapper = shallow(<Button size="large" />)
    expect(wrapper.find('.btn.btn-lg').length).toEqual(1)
  })

  test('should render any passed className', () => {
    const wrapper = shallow(<Button className="special classes" />)
    expect(wrapper.find('.btn.special.classes').length).toEqual(1)
  })

  test('should render any other passed attributes', () => {
    const wrapper = shallow(<Button data-test="totally-rad" />)
    expect(wrapper.find('[data-test="totally-rad"]').length).toEqual(1)
  })

  test('should not render class btn-anchor without passed prop', () => {
    const wrapper = shallow(<Button />)
    expect(wrapper.find('button.btn-anchor').length).toEqual(0)
  })

  test('should render class btn-anchor when passed prop', () => {
    const wrapper = shallow(<Button link />)
    expect(wrapper.find('button.btn-anchor').length).toEqual(1)
  })

  test('should render a passed type', () => {
    const wrapper = shallow(<Button type="reset" />)
    expect(wrapper.find('button[type="reset"]').length).toEqual(1)
  })

  test('should render children', () => {
    const wrapper = shallow(
      <Button>
        <span>Rad</span>
      </Button>
    )
    expect(wrapper.find('span').length).toEqual(1)
  })

  test('simulates click events', () => {
    const onButtonClick = jest.fn()
    const wrapper = shallow(<Button onClick={onButtonClick} />)
    wrapper.find('button').simulate('click')
    expect(onButtonClick).toHaveBeenCalled()
  })

  // Check that outline blur handler doesn't interfere with passing `onMouseDown`
  test('attaches onMouseDown events correctly', () => {
    const onMouseDown = jest.fn()
    const wrapper = shallow(<Button onMouseDown={onMouseDown} />)

    // Pass a mocked `evt` object to `simulate` with required props for
    // suppressClickOutline handler
    wrapper.find('button').simulate('mousedown', {
      target: {
        style: { outline: '' },
        addEventListener: () => {},
        removeEventListener: () => {}
      }
    })

    expect(onMouseDown).toHaveBeenCalled()
  })
})
