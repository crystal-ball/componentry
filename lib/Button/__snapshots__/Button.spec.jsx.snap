// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots it renders brand color correctly 1`] = `
<button
  className="btn btn-success"
  type="button"
>
  Facebook
</button>
`;

exports[`<Button /> Snapshots it renders defaults correctly 1`] = `
<button
  className="btn btn-primary"
  type="button"
>
  Facebook
</button>
`;

exports[`<Button /> Snapshots it renders large outline correctly 1`] = `
<button
  className="btn btn-outline-success btn-lg"
  type="button"
>
  Facebook
</button>
`;

exports[`<Button /> Snapshots it renders link style correctly 1`] = `
<button
  className="btn btn-anchor"
  type="button"
>
  Facebook
</button>
`;

exports[`<Button /> Snapshots it renders outline correctly 1`] = `
<button
  className="btn btn-outline-success"
  type="button"
>
  Facebook
</button>
`;
